Source/WebKit/Shared/Extensions/_WKWebExtensionLocalization.mm

```diff
@@ -0,0 +1,334 @@
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#if !__has_feature(objc_arc)
+#error This file requires ARC. Add the "-fobjc-arc" compiler flag for this file.
+#endif
+
+#import "config.h"
+#import "_WKWebExtensionLocalization.h"
+
+#import "CocoaHelpers.h"
+#import "Logging.h"
+
+NSString * const WKLocalizationDictionaryMessageKey = @"message";
+NSString * const WKLocalizationDictionaryDescriptionKey = @"description";
+NSString * const WKLocalizationDictionaryPlaceholdersKey = @"placeholders";
+NSString * const WKLocalizationDictionaryJSONPathFormat = @"_locales/%@/messages.json";
+
+NSString * const WKPredefinedMessageUILocale = @"@@ui_locale";
+NSString * const WKPredefinedMessageLanguageDirection = @"@@bidi_dir";
+NSString * const WKPredefinedMessageLanguageDirectionReversed = @"@@bidi_reversed_dir";
+NSString * const WKPredefinedMessageTextLeadingEdge = @"@@bidi_start_edge";
+NSString * const WKPredefinedMessageTextTrailingEdge = @"@@bidi_end_edge";
+NSString * const WKPredefinedMessageExtensionID = @"@@extension_id";
+
+NSString * const WKPredefinedMessageValueLeftToRight = @"ltr";
+NSString * const WKPredefinedMessageValueRightToLeft = @"rtl";
+NSString * const WKPredefinedMessageValueTextEdgeLeft = @"left";
+NSString * const WKPredefinedMessageValueTextEdgeRight = @"right";
+
+static NSString * const placeholderDictionaryContentKey = @"content";
+
+// NSCoding keys
+static NSString * const localizationDictionaryCodingKey = @"localizationDictionary";
+static NSString * const localeStringDictionaryCodingKey = @"localeString";
+static NSString * const localeCodingKey = @"locale";
+static NSString * const uniqueIdentifierCodingKey = @"uniqueIdentifier";
+
+typedef NSDictionary<NSString *, NSDictionary *> LocalizationDictionary;
+typedef NSDictionary<NSString *, NSDictionary<NSString *, NSString *> *> PlaceholderDictionary;
+
+using namespace WebKit;
+
+static LocalizationDictionary *localizationDictionaryAtURL(NSURL *dictionaryFileURL)
+{
+    NSData *dictionaryData = [NSData dataWithContentsOfURL:dictionaryFileURL];
+    if (!dictionaryData)
+        return nil;
+
+    NSError *jsonParsingError;
+    LocalizationDictionary *localizationDictionary = (NSDictionary *)[NSJSONSerialization JSONObjectWithData:dictionaryData options:0 error:&jsonParsingError];
```

Comment:
This should use `dynamic_objc_cast<NSDictionary>()` instead of `(NSDictionary *)`, since it isn't guaranteed to be a dictionary.